File: src/main/java/fp/yeyu/tos/SpiritEntityHeadFeatureRenderer.java
package fp.yeyu.tos;

import com.mojang.authlib.GameProfile;
import com.mojang.authlib.minecraft.MinecraftProfileTexture;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.render.OverlayTexture;
import net.minecraft.client.render.RenderLayer;
import net.minecraft.client.render.VertexConsumer;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.render.entity.feature.FeatureRenderer;
import net.minecraft.client.render.entity.feature.FeatureRendererContext;
import net.minecraft.client.render.entity.model.SkullEntityModel;
import net.minecraft.client.render.entity.model.SkullOverlayEntityModel;
import net.minecraft.client.util.DefaultSkinHelper;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.entity.player.PlayerEntity;

import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public class SpiritEntityHeadFeatureRenderer extends FeatureRenderer<SpiritEntity, SpiritEntityModel> {

    private static final SkullEntityModel skullEntityModel = new SkullOverlayEntityModel();
    private static final GameProfile DUMMY_GP = new GameProfile(new UUID(1, 0), "Steve");


    public SpiritEntityHeadFeatureRenderer(FeatureRendererContext<SpiritEntity, SpiritEntityModel> context) {
        super(context);
    }

    @Override
    public void render(MatrixStack matrixStack, VertexConsumerProvider vertexConsumerProvider, int light, SpiritEntity entity, float limbAngle, float limbDistance, float tickDelta, float animationProgress, float headYaw, float headPitch) {
        matrixStack.push();
        this.getContextModel().getHead().rotate(matrixStack);

        matrixStack.scale(0.9F, -0.9F, -0.9F);
        final GameProfile gameProfileLocal = Optional.ofNullable(MinecraftClient.getInstance()).map(e -> e.player).map(PlayerEntity::getGameProfile).orElse(DUMMY_GP);
        final GameProfile gameProfile = Optional.ofNullable(entity.closestPlayer).map(PlayerEntity::getGameProfile).orElse(gameProfileLocal);

        matrixStack.translate(-0.5D, -0.23D, -0.5D);
        render(180.0F, gameProfile, entity.pitch, matrixStack, vertexConsumerProvider, light, entity.getHeadAlphaLevelPlayer().next().floatValue());
        matrixStack.pop();
    }

    public static void render(float limbAngle, GameProfile gameProfile, float g, MatrixStack matrixStack, VertexConsumerProvider vertexConsumerProvider, int light, float alpha) {
        matrixStack.push();
        matrixStack.translate(0.5D, 0.0D, 0.5D);
        matrixStack.scale(-1.0F, -1.0F, 1.0F);
        VertexConsumer vertexConsumer = vertexConsumerProvider.getBuffer(getTransparencyRenderLayer(gameProfile));
        skullEntityModel.method_2821(g, limbAngle, 0.0F);
        skullEntityModel.render(matrixStack, vertexConsumer, light, OverlayTexture.DEFAULT_UV, 1, 1, 1, alpha);
        matrixStack.pop();
    }

    private static RenderLayer getTransparencyRenderLayer(GameProfile gameProfile) {
        MinecraftClient minecraftClient = MinecraftClient.getInstance();
        Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> map = minecraftClient.getSkinProvider().getTextures(gameProfile);
        return map.containsKey(MinecraftProfileTexture.Type.SKIN) ?
                RenderLayer.getEntityTranslucent(minecraftClient.getSkinProvider().loadSkin(map.get(MinecraftProfileTexture.Type.SKIN), MinecraftProfileTexture.Type.SKIN)) :
                RenderLayer.getEntityTranslucent(DefaultSkinHelper.getTexture(PlayerEntity.getUuidFromProfile(gameProfile)));
    }

}

File: src/main/kotlin/fp/yeyu/tos/client/SpiritEntityHeadFeatureRenderer.kt
package fp.yeyu.tos.client

import com.mojang.authlib.GameProfile
import com.mojang.authlib.minecraft.MinecraftProfileTexture
import fp.yeyu.tos.entity.SpiritEntity
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.OverlayTexture
import net.minecraft.client.render.RenderLayer
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.entity.feature.FeatureRenderer
import net.minecraft.client.render.entity.feature.FeatureRendererContext
import net.minecraft.client.render.entity.model.SkullEntityModel
import net.minecraft.client.render.entity.model.SkullOverlayEntityModel
import net.minecraft.client.util.DefaultSkinHelper
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.entity.player.PlayerEntity
import java.util.*

class SpiritEntityHeadFeatureRenderer(context: FeatureRendererContext<SpiritEntity?, SpiritEntityModel?>?) : FeatureRenderer<SpiritEntity, SpiritEntityModel?>(context) {
    override fun render(matrixStack: MatrixStack, vertexConsumerProvider: VertexConsumerProvider, light: Int, entity: SpiritEntity, limbAngle: Float, limbDistance: Float, tickDelta: Float, animationProgress: Float, headYaw: Float, headPitch: Float) {
        matrixStack.push()
        val contextModel = this.contextModel ?: return
        contextModel.head.rotate(matrixStack)
        matrixStack.scale(0.9f, -0.9f, -0.9f)
        val gameProfile = entity.closestPlayer?.gameProfile ?: MinecraftClient.getInstance()?.player?.gameProfile ?: dummyGameProfile
        matrixStack.translate(-0.5, -0.23, -0.5)
        render(gameProfile, entity.pitch, matrixStack, vertexConsumerProvider, light, entity.headAlphaLevelPlayer.next().toFloat())
        matrixStack.pop()
    }

    private val skullEntityModel: SkullEntityModel = SkullOverlayEntityModel()
    private val dummyGameProfile = GameProfile(UUID(1, 0), "Steve")

    private fun render(gameProfile: GameProfile, limbAngle: Float, matrixStack: MatrixStack, vertexConsumerProvider: VertexConsumerProvider, light: Int, alpha: Float) {
        matrixStack.push()
        matrixStack.translate(0.5, 0.0, 0.5)
        matrixStack.scale(-1.0f, -1.0f, 1.0f)
        val vertexConsumer = vertexConsumerProvider.getBuffer(getTransparencyRenderLayer(gameProfile))
        setAngle(limbAngle, 180f, 0.0f)
        skullEntityModel.render(matrixStack, vertexConsumer, light, OverlayTexture.DEFAULT_UV, 1f, 1f, 1f, alpha)
        matrixStack.pop()
    }

    private fun getTransparencyRenderLayer(gameProfile: GameProfile): RenderLayer {
        val minecraftClient = MinecraftClient.getInstance()
        val map = minecraftClient.skinProvider.getTextures(gameProfile)
        return if (map.containsKey(MinecraftProfileTexture.Type.SKIN)) RenderLayer.getEntityTranslucent(minecraftClient.skinProvider.loadSkin(map[MinecraftProfileTexture.Type.SKIN], MinecraftProfileTexture.Type.SKIN)) else RenderLayer.getEntityTranslucent(DefaultSkinHelper.getTexture(PlayerEntity.getUuidFromProfile(gameProfile)))
    }

    private fun setAngle(factor: Float, from: Float, to: Float) {
        skullEntityModel.method_2821(factor, from, to)
    }
}